File: src/cookbook/expect.js
// expect.js — what a step asserts, evaluated. The JS half of the expectation
// vocabulary in `kernel/src/scenario.rs`; the two are pinned to identical
// answers by test/cookbook.test.js.
//
// `check` returns every expectation that did not hold, in the corpus's own
// words, plus the values it actually saw. The count of keys is returned
// separately because "nothing failed" and "nothing was asserted" are different
// states, and a runner that conflates them reports green for the worst reason.
import { at, lenOf, show, typeName } from "./tokens.js";

export const KEYS = ["status", "status_in", "max_ms", "json", "json_not", "json_in", "json_type",
  "json_present", "json_absent", "json_len_at_least", "json_len_at_most", "json_gte", "json_lte",
  "json_matches", "each", "contains", "not_both",
  "rc", "stdout_contains", "stderr_contains", "contains_text", "absent_text", "matches"];

const eq = (a, b) => (typeof a === "number" && typeof b === "number" ? Math.abs(a - b) < 1e-9 : JSON.stringify(a) === JSON.stringify(b));
const numOf = (v) => (typeof v === "number" ? v : typeof v === "string" && v.trim() !== "" && Number.isFinite(Number(v)) ? Number(v) : null);
const pairs = (o) => (o && typeof o === "object" && !Array.isArray(o) ? Object.entries(o) : []);

/** How many expectation keys this block carries, and which are not implemented. */
export function asserts(expect) {
  const keys = Object.keys(expect || {});
  return { n: keys.filter((k) => KEYS.includes(k)).length, unknown: keys.filter((k) => !KEYS.includes(k)) };
}

export function check(expect, body, status, ms) {
  const why = [];
  const got = {};
  // `got` is evidence a person reads, not the response: a 900-row list under
  // one path would push the actual failure off the board.
  const note = (p, v) => { const t = show(v === undefined ? null : v); got[p] = t != null && t.length > 240 ? `${t.slice(0, 240)}… (${t.length} bytes)` : (v === undefined ? null : v); };
  const e = expect || {};

  if (typeof e.status === "number" && status !== e.status) why.push(`status ${status}, expected ${e.status}`);
  if (Array.isArray(e.status_in) && !e.status_in.includes(status)) why.push(`status ${status}, expected one of ${JSON.stringify(e.status_in)}`);
  if (typeof e.max_ms === "number" && ms > e.max_ms) why.push(`took ${Math.round(ms)}ms, budget ${e.max_ms}ms`);

  for (const [p, want] of pairs(e.json)) {
    const have = at(body, p); note(p, have);
    if (have === undefined) why.push(`${p} is absent, expected ${show(want)}`);
    else if (!eq(have, want)) why.push(`${p} = ${show(have)}, expected ${show(want)}`);
  }
  for (const [p, want] of pairs(e.json_not)) {
    const have = at(body, p); note(p, have);
    if (have !== undefined && eq(have, want)) why.push(`${p} = ${show(have)}, expected anything else`);
  }
  for (const [p, set] of pairs(e.json_in)) {
    const have = at(body, p); note(p, have);
    if (!Array.isArray(set) || !set.some((x) => eq(x, have))) why.push(`${p} = ${have === undefined ? "absent" : show(have)}, expected one of ${JSON.stringify(set)}`);
  }
  for (const [p, want] of pairs(e.json_type)) {
    const have = at(body, p); note(p, have);
    const t = have === undefined ? "absent" : typeName(have);
    const ok = t === want || ((want === "number" || want === "float") && typeof have === "number");
    if (!ok) why.push(`${p} is ${t}, expected ${want}`);
  }
  for (const p of Array.isArray(e.json_present) ? e.json_present : []) {
    const have = at(body, p); note(p, have);
    if (have === undefined || have === null) why.push(`${p} is absent or null`);
  }
  for (const p of Array.isArray(e.json_absent) ? e.json_absent : []) {
    const have = at(body, p); note(p, have);
    if (have !== undefined && have !== null) why.push(`${p} is present (${show(have)}), expected absent`);
  }
  for (const [p, want] of pairs(e.json_len_at_least)) {
    const have = at(body, p); note(p, have);
    const n = lenOf(have), w = numOf(want);
    if (n == null || w == null) why.push(`${p} has no length`);
    else if (n < w) why.push(`${p} has ${n} items, expected at least ${w}`);
  }
  for (const [p, want] of pairs(e.json_len_at_most)) {
    const have = at(body, p); note(p, have);
    const n = lenOf(have), w = numOf(want);
    if (n == null || w == null) why.push(`${p} has no length`);
    else if (n > w) why.push(`${p} has ${n} items, expected at most ${w}`);
  }
  for (const [key, op, label] of [["json_gte", (a, b) => a >= b, ">="], ["json_lte", (a, b) => a <= b, "<="]]) {
    for (const [p, want] of pairs(e[key])) {
      const have = at(body, p); note(p, have);
      const v = numOf(have), w = numOf(want);
      if (v == null || w == null) why.push(`${p} is not a number`);
      else if (!op(v, w)) why.push(`${p} = ${v}, expected ${label} ${w}`);
    }
  }
  for (const [p, pat] of pairs(e.json_matches)) {
    const have = at(body, p); note(p, have);
    const subject = have === undefined ? "" : show(have);
    let re = null;
    try { re = new RegExp(String(pat)); } catch (err) { why.push(`${p}: /${pat}/ is not a valid pattern (${err.message})`); }
    if (re && !re.test(subject)) why.push(`${p} = ${JSON.stringify(subject)}, expected to match /${pat}/`);
  }
  for (const [p, nested] of pairs(e.each)) {
    const have = at(body, p);
    if (!Array.isArray(have)) { why.push(`${p} is ${have === undefined ? "absent" : typeName(have)}, expected a list to iterate`); continue; }
    have.forEach((item, i) => { for (const w of check(nested, item, status, ms).why) why.push(`${p}[${i}]: ${w}`); });
  }
  for (const [p, want] of pairs(e.contains)) {
    const have = at(body, p);
    if (!Array.isArray(have)) { why.push(`${p} is ${have === undefined ? "absent" : typeName(have)}, expected a list`); continue; }
    if (!have.some((item) => pairs(want).every(([f, v]) => eq(at(item, f), v)))) why.push(`no item of ${p} (${have.length} of them) matches ${JSON.stringify(want)}`);
  }
  if (Array.isArray(e.not_both)) {
    if (e.not_both.length !== 2) why.push("not_both takes exactly two blocks");
    else {
      const holds = (side) => pairs(side).every(([p, v]) => eq(at(body, p), v));
      if (holds(e.not_both[0]) && holds(e.not_both[1])) why.push(`both held and they contradict: ${JSON.stringify(e.not_both[0])} AND ${JSON.stringify(e.not_both[1])}`);
    }
  }
  return { why, got };
}

/** The `run` and `static` halves, which share the vocabulary but not the body. */
/** The expectation keys that read a BODY rather than a stream. A command whose
 *  stdout is JSON — `bb dotty shot --json`, `bb findings --json`, anything with
 *  a `--json` flag — should be assertable the way a response is, instead of
 *  through `stdout_contains` against a serialised object, which breaks on key
 *  order and reads as slop. */
export const BODY_KEYS = KEYS.filter((k) => k.startsWith("json") || ["each", "contains", "not_both"].includes(k));

/** stdout as a body, or null when it is not one. Only an object or an array
 *  counts: a command printing the bare word `true` has not returned a body, and
 *  treating it as one would make `json` assertions pass against nothing. */
export function stdoutBody(stdout) {
  const t = String(stdout || "").trim();
  if (!t || !/^[[{]/.test(t)) return null;
  try { const v = JSON.parse(t); return v && typeof v === "object" ? v : null; }
  catch { return null; }
}

export function checkCmd(expect, { rc, stdout, stderr, ms }) {
  const why = [];
  let n = 0;
  const e = expect || null;
  if (e && typeof e.rc === "number") { n++; if (rc !== e.rc) why.push(`rc ${rc}, expected ${e.rc}`); }
  for (const [key, hay, label] of [["stdout_contains", stdout, "stdout"], ["stderr_contains", stderr, "stderr"]]) {
    const want = e && e[key];
    for (const needle of typeof want === "string" ? [want] : Array.isArray(want) ? want : []) {
      n++;
      if (!String(hay).includes(needle)) why.push(`${label} does not contain ${JSON.stringify(needle)}`);
    }
  }
  if (e && typeof e.max_ms === "number") { n++; if (ms > e.max_ms) why.push(`took ${Math.round(ms)}ms, budget ${e.max_ms}ms`); }

  // Body assertions against stdout. The sub-block carries only body keys, so
  // `status` and `max_ms` cannot fire twice — `rc` and `max_ms` are the
  // command's own and are handled above.
  const bodyWanted = e ? BODY_KEYS.filter((k) => e[k] !== undefined) : [];
  if (bodyWanted.length) {
    const body = stdoutBody(stdout);
    if (body === null) {
      n += bodyWanted.length;
      why.push(`stdout is not a JSON object or array, so ${bodyWanted.join(", ")} could not be checked (add --json to the command?)`);
    } else {
      const sub = Object.fromEntries(bodyWanted.map((k) => [k, e[k]]));
      const r = check(sub, body, rc, ms);
      n += asserts(sub).n;
      why.push(...r.why);
    }
  }

  // A bare `run` with no expect block means "this must succeed" — otherwise it
  // is a step that ran a command and checked nothing.
  if (n === 0 && !e) { n = 1; if (rc !== 0) why.push(`rc ${rc} and nothing was asserted; a bare \`run\` step expects 0`); }
  return { why, n };
}

// ── the ui half ─────────────────────────────────────────────────────────────
//
// The fourth step kind, and the one thing about it that matters: there is no
// model between the step and the action. `{"ui": "click #submit"}` is a line an
// actor wrote ONCE, at practice time, and every replay after that costs
// nothing. A driver that took natural language per step would be a model call
// per step — ten minutes and twenty-four thousand tokens per drive, measured in
// `src/recom/artemis.js` — against a corpus this factory replays on every tick.
//
// The grammar is four verbs and nothing else. A fifth would be a place for a
// sentence to hide.
export const UI_VERBS = ["open", "click", "type", "expect"];

/** `{ verb, selector, text }` from a ui line, or `{ why }`. */
export function parseUi(line) {
  const raw = String(line || "").trim();
  if (!raw) return { why: "`ui` is empty" };
  const [head, ...rest] = raw.split(/\s+/);
  const verb = head.toLowerCase();
  if (!UI_VERBS.includes(verb)) return { why: `\`${head}\` is not a ui verb; one of ${UI_VERBS.join(", ")}` };
  const tail = rest.join(" ").trim();
  if (verb === "open") return tail ? { verb, target: tail } : { why: "`open` needs a page: `open index.html`" };
  if (verb === "click") return tail ? { verb, selector: tail } : { why: "`click` needs a selector: `click #submit`" };
  if (verb === "type") {
    const m = /^(\S+)\s+(.+)$/.exec(tail);
    return m ? { verb, selector: m[1], text: unquote(m[2]) } : { why: '`type` needs a selector and a value: `type #email a@b.c`' };
  }
  // expect: `expect text "..."` | `expect visible <selector>` | `expect absent <selector>`
  const m = /^(text|visible|absent)\s+(.+)$/.exec(tail);
  if (!m) return { why: '`expect` takes `text <string>`, `visible <selector>` or `absent <selector>`' };
  return m[1] === "text" ? { verb, what: "text", text: unquote(m[2]) } : { verb, what: m[1], selector: m[2].trim() };
}

const unquote = (s) => { const t = String(s).trim(); return /^(".*"|'.*')$/s.test(t) ? t.slice(1, -1) : t; };

/** Does a ui action assert anything? `click` and `type` drive; only `expect`
 *  makes a claim, and a scenario of nothing but drives is green because nothing
 *  was checked — the state `corpus.check()` exists to refuse. */
export const uiAsserts = (action) => action.verb === "expect";

File: kernel/src/scenario/expect.rs
//! scenario/expect.rs — what a step asserted, and what the system actually did.
//!
//! One function decides every red step in the factory, so it lives on its own:
//! a change to how `json_matches` compares must not be made in the same file as
//! a change to how a socket is opened. `asserts` is the other half of the same
//! rule — it counts what a step checks, and a step that checks nothing is
//! reported `empty` rather than green.
use crate::json::Json;
use crate::rx;
use crate::subst::{at, len_of, show, type_name};

pub fn each_pair(v: Option<&Json>, mut f: impl FnMut(&str, &Json)) {
    if let Some(Json::Obj(m)) = v { for (k, val) in m { f(k, val); } }
}

pub fn eq(a: &Json, b: &Json) -> bool {
    match (a, b) { (Json::Num(x), Json::Num(y)) => (x - y).abs() < 1e-9, _ => a == b }
}

pub fn num_of(v: &Json) -> Option<f64> { match v { Json::Num(n) => Some(*n), Json::Str(s) => s.parse().ok(), _ => None } }

/// Every expectation that did not hold, in the corpus's own words. An empty
/// result and "nothing was asserted" are different states and the caller
/// separates them: `asserts()` counts the keys.
pub fn check(expect: &Json, body: &Json, status: u16, ms: f64, got: &mut Json) -> Vec<String> {
    let mut why = Vec::new();
    // `got` is evidence a person reads, not the response: a 900-row list under
    // one path would push the actual failure off the board.
    let mut note = |path: &str, v: Option<&Json>| {
        let val = v.cloned().unwrap_or(Json::Null);
        let txt = val.to_string();
        got.set(path, if txt.len() > 240 { Json::Str(format!("{}… ({} bytes)", &txt[..240.min(txt.len())], txt.len())) } else { val });
    };
    if let Some(Json::Num(want)) = expect.get("status") {
        if status as f64 != *want { why.push(format!("status {}, expected {}", status, *want as i64)); }
    }
    if let Some(Json::Arr(a)) = expect.get("status_in") {
        if !a.iter().any(|x| x.as_f64().map(|n| n as u16 == status).unwrap_or(false)) {
            why.push(format!("status {}, expected one of {}", status, Json::Arr(a.clone()).to_string()));
        }
    }
    if let Some(Json::Num(b)) = expect.get("max_ms") {
        if ms > *b { why.push(format!("took {:.0}ms, budget {:.0}ms", ms, b)); }
    }
    each_pair(expect.get("json"), |p, want| {
        let have = at(body, p);
        note(p, have);
        match have { Some(v) if eq(v, want) => {}, Some(v) => why.push(format!("{} = {}, expected {}", p, show(v), show(want))), None => why.push(format!("{} is absent, expected {}", p, show(want))) }
    });
    each_pair(expect.get("json_not"), |p, want| {
        let have = at(body, p);
        note(p, have);
        if let Some(v) = have { if eq(v, want) { why.push(format!("{} = {}, expected anything else", p, show(v))); } }
    });
    each_pair(expect.get("json_in"), |p, set| {
        let have = at(body, p);
        note(p, have);
        let ok = match (have, set) { (Some(v), Json::Arr(a)) => a.iter().any(|x| eq(x, v)), _ => false };
        if !ok { why.push(format!("{} = {}, expected one of {}", p, have.map(show).unwrap_or("absent".into()), set.to_string())); }
    });
    each_pair(expect.get("json_type"), |p, want| {
        let have = at(body, p);
        note(p, have);
        let want_s = show(want);
        let ok = match have {
            Some(v) => { let t = type_name(v); t == want_s || (want_s == "number" && matches!(v, Json::Num(_))) || (want_s == "float" && matches!(v, Json::Num(_))) }
            None => false,
        };
        if !ok { why.push(format!("{} is {}, expected {}", p, have.map(|v| type_name(v).to_string()).unwrap_or("absent".into()), want_s)); }
    });
    if let Some(Json::Arr(a)) = expect.get("json_present") {
        for p in a.iter().filter_map(|x| x.as_str()) {
            let have = at(body, p);
            note(p, have);
            if !matches!(have, Some(v) if !matches!(v, Json::Null)) { why.push(format!("{} is absent or null", p)); }
        }
    }
    if let Some(Json::Arr(a)) = expect.get("json_absent") {
        for p in a.iter().filter_map(|x| x.as_str()) {
            let have = at(body, p);
            note(p, have);
            if matches!(have, Some(v) if !matches!(v, Json::Null)) { why.push(format!("{} is present ({}), expected absent", p, show(have.unwrap()))); }
        }
    }
    each_pair(expect.get("json_len_at_least"), |p, want| {
        let have = at(body, p);
        note(p, have);
        let n = have.and_then(len_of);
        match (n, num_of(want)) { (Some(n), Some(w)) if (n as f64) >= w => {}, (Some(n), Some(w)) => why.push(format!("{} has {} items, expected at least {}", p, n, w as i64)), _ => why.push(format!("{} has no length", p)) }
    });
    each_pair(expect.get("json_len_at_most"), |p, want| {
        let have = at(body, p);
        note(p, have);
        let n = have.and_then(len_of);
        match (n, num_of(want)) { (Some(n), Some(w)) if (n as f64) <= w => {}, (Some(n), Some(w)) => why.push(format!("{} has {} items, expected at most {}", p, n, w as i64)), _ => why.push(format!("{} has no length", p)) }
    });
    each_pair(expect.get("json_gte"), |p, want| {
        let have = at(body, p);
        note(p, have);
        match (have.and_then(num_of), num_of(want)) { (Some(v), Some(w)) if v >= w => {}, (Some(v), Some(w)) => why.push(format!("{} = {}, expected >= {}", p, v, w)), _ => why.push(format!("{} is not a number", p)) }
    });
    each_pair(expect.get("json_lte"), |p, want| {
        let have = at(body, p);
        note(p, have);
        match (have.and_then(num_of), num_of(want)) { (Some(v), Some(w)) if v <= w => {}, (Some(v), Some(w)) => why.push(format!("{} = {}, expected <= {}", p, v, w)), _ => why.push(format!("{} is not a number", p)) }
    });
    each_pair(expect.get("json_matches"), |p, pat| {
        let have = at(body, p);
        note(p, have);
        let subject = have.map(show).unwrap_or_default();
        match rx::test(&show(pat), &subject) {
            Ok(true) => {}
            Ok(false) => why.push(format!("{} = {:?}, expected to match /{}/", p, subject, show(pat))),
            Err(e) => why.push(format!("{}: pattern /{}/ is outside the kernel's subset ({}) — run this corpus with --engine js", p, show(pat), e)),
        }
    });
    each_pair(expect.get("each"), |p, nested| {
        match at(body, p) {
            Some(Json::Arr(items)) => {
                for (i, item) in items.iter().enumerate() {
                    let mut sub = Json::obj();
                    for w in check(nested, item, status, ms, &mut sub) { why.push(format!("{}[{}]: {}", p, i, w)); }
                }
            }
            Some(v) => why.push(format!("{} is {}, expected a list to iterate", p, type_name(v))),
            None => why.push(format!("{} is absent, expected a list to iterate", p)),
        }
    });
    each_pair(expect.get("contains"), |p, want| {
        match at(body, p) {
            Some(Json::Arr(items)) => {
                let hit = items.iter().any(|item| {
                    let mut all = true;
                    each_pair(Some(want), |f, v| { if !at(item, f).map(|x| eq(x, v)).unwrap_or(false) { all = false; } });
                    all
                });
                if !hit { why.push(format!("no item of {} ({} of them) matches {}", p, items.len(), want.to_string())); }
            }
            Some(v) => why.push(format!("{} is {}, expected a list", p, type_name(v))),
            None => why.push(format!("{} is absent, expected a list", p)),
        }
    });
    if let Some(Json::Arr(sides)) = expect.get("not_both") {
        if sides.len() == 2 {
            let holds = |side: &Json| { let mut all = true; each_pair(Some(side), |p, v| { if !at(body, p).map(|x| eq(x, v)).unwrap_or(false) { all = false; } }); all };
            if holds(&sides[0]) && holds(&sides[1]) {
                why.push(format!("both held and they contradict: {} AND {}", sides[0].to_string(), sides[1].to_string()));
            }
        } else { why.push("not_both takes exactly two blocks".into()); }
    }
    why
}

const KEYS: &[&str] = &["status", "status_in", "max_ms", "json", "json_not", "json_in", "json_type", "json_present", "json_absent",
    "json_len_at_least", "json_len_at_most", "json_gte", "json_lte", "json_matches", "each", "contains", "not_both",
    "rc", "stdout_contains", "stderr_contains", "contains_text", "absent_text", "matches"];

/// How many expectation keys a block actually carries, and which keys it used
/// that this runner does not implement. A key nobody implements is silently
/// green, which is the failure mode `check` exists to prevent.
pub fn asserts(expect: &Json) -> (usize, Vec<String>) {
    let mut n = 0;
    let mut unknown = Vec::new();
    if let Json::Obj(m) = expect {
        for k in m.keys() { if KEYS.contains(&k.as_str()) { n += 1; } else { unknown.push(k.clone()); } }
    }
    (n, unknown)
}
